libxml.3
.TH libxml 3 "12 April 2000"
.SH NAME
libxml \- library used to parse XML files
.SH DESCRIPTION
The
.I  libxml
library is used to parse XML files. 
Its internal document repesentation is as close as possible to the 
.I DOM 
(Document Object Model) interface,
an API for accessing XML or HTML structured documents.
.LP
The
.I libxml
library also has a 
.IR SAX -like
interface, 
which is designed to be compatible with 
.IR expat (1).
NOTE:
.IR SAX , 
the Simple API for XML,
is a standard interface for event-based XML parsing,
developed collaboratively by the members of the XML-DEV mailing list, 
currently hosted by OASIS.
The
.I expat
library is a XML 1.0 parser written in C,
which aims to be fully conforming. 
It is currently not a validating XML processor.
.LP
The
.I libxml 
library now includes a nearly complete 
.I XPath 
implementation. 
The
.I XPath
(XML Path Language) is a language for addressing parts of an 
XML document,
designed to be used by both 
.I XSLT 
and 
.IR XPointer .
.LP
The
.I libxml 
library exports Push and Pull type parser interfaces for both XML and 
.IR html . 
.SH FILES
.TP 2.2i
.B /depot/lib/libxml_2.0.0/libxml.a
static library
.TP
.B /depot/lib/libxml_2.0.0/libxml.so
shareable library
.TP
.B /depot/package/libxml_2.0.0/bin/xmllint
binary application for parsing XML files
.SH AUTHORS
Daniel Veillard (daniel@veillard.com).
If you download and install this package please send the author email.
Manual page by Ziying Sherwin (sherwin@nlm.nih.gov),
Lister Hill National Center for Biomedical Communications,
U.S. National Library of Medicine.
.SH SEE ALSO
.IR xmllint (1),
.IR libxslt (3),
.IR libexslt (3),
.IR xsltproc (1)
.\" end of manual page


NEWS
NEWS file for libxml2

v2.14.4: Jun 12 2025

### Regressions

- parser: Fix parsing of PublicIds and VersionNums
- parser: Fix custom SAX parsers without cdataBlock handler
- error: Fix initGenericErrorDefaultFunc compatibility macro again
- io: Make xmlOutputBufferCreate* not free encoder on error
- reader: Fix null deref on malloc failure
- Revert "meson: Install libxml2.py"

### Security

- tree: Fix integer overflow in xmlBuildQName

### Improvements

- parser: Use parser context as default in resource loader
- parser: Only validate EnumerationTypes when requested
- parser: Undeprecate some parser context members

### Build systems

- cmake: Avoid overlinking with non-CMake libxml2-config.cmake
- cmake: Make iconv a private dependency


v2.14.3: May 13 2025

### Regressions

- reader: Fix reading compressed data
- parser: Make undeclared entities in XML content fatal
- save: Fix XML escape table
- save: Fix xmlSave with NULL encoding
- Revert "valid: Remove duplicate error messages when streaming"

### Bug fixes

- save: Fix serialization of attribute defaults containing &lt;
- io: Fix linkage of __xml*BufferCreateFilename functions

### Build systems

- cmake: Fix installation directories in libxml2-config.cmake
- meson: Install libxml2.py

### Improvements

- parser: Make xmlCtxtGetValidCtxt depend on VALID_ENABLED
- html: Avoid HTML_PARSE_HTML5 clashing with XML_PARSE_NOENT


v2.14.2: Apr 17 2025

### Security

- [CVE-2025-32415] schemas: Fix heap buffer overflow in
  xmlSchemaIDCFillNodeTables
- [CVE-2025-32414] python: Read at most len/4 characters. (Maks Verver)

### Build

- error: Fix initGenericErrorDefaultFunc compatibility macro
- meson: don't link with pthreads on Windows (Benjamin Gilbert)
- cmake, meson: Align Darwin version info with Autotools
- globals: Fix --with-thread-alloc build
- meson: ensure relaxng option supports minimum option (Lovell Fuller)


v2.14.1: Apr 3 2025

### Regressions

- parser: Fix XML_PARSE_NOBLANKS dropping non-whitespace text

### Build systems

- win32-legacy: Fix build (ThomasK)
- meson: Fix build from tarball
- cmake, meson: Change library filename to libxml2.so.16.0.0


v2.14.0: Mar 27 2025

### Major changes

The HTML tokenizer now conforms fully to HTML5. Several non-standard
syntax warnings were removed. Note that HTML5 tree construction isn't
implemented yet.

Binary compatibility is restricted to versions 2.14 or newer. On ELF
systems, the soname was bumped from libxml2.so.2 to libxml2.so.16.

The serialization API will now take user-provided or default encodings
into account when serializing attribute values, matching the
serialization of text and avoiding unnecessary escaping.

The XML parser won't try to merge consecutive CDATA sections as before
to align with web standards. Each CDATA section will create exactly one
node or SAX callback.

Support for RELAX NG can now be disabled with a new configuration
option independently of XML Schemas support. It is still enabled by
default.

The "legacy" configuration option won't enable support for HTTP and
LZMA anymore. These features will be removed in the next release.

Parts of the xmllint executable were refactored, allowing the
combination of more options. OOM errors should be reported reliably now.

Several improvements were made to the build systems. Meson is fully
supported now.

Parts of the buffering code were reworked and simplified.

Overflow checks before reallocations were hardenend.

Some unprefixed symbols were renamed to avoid namespace pollution.

### Other potentially incompatible changes

Strings passed to the "characters" callback of the HTML SAX parser
aren't null-terminated anymore, matching the behavior of the XML
parser. Custom SAX parsers must use the "len" argument.

xmlIOParseDTD doesn't allow null bytes at the end of the input anymore.

Type and layout of conversion callbacks in struct xmlCharEncodingHandler
were changed. Applications using libxml2's encoding conversion API
should use functions xmlCharEncInFunc and xmlCharEncOutFunc instead of
accessing the callbacks directly.

### New features

Input callbacks can now be set on a parser context and an improved API
to create parser input is available. The following new functions,
taking a parser input object, were added:

- xmlCtxtParseDocument
- xmlCtxtParseContent as replacement for xmlParseBalancedChunkMemory
  and xmlParseInNodeContext
- xmlCtxtParseDtd

The xmlSave API now has additional options to replace global settings.

Parser options XML_PARSE_UNZIP, XML_PARSE_NO_SYS_CATALOG and
XML_PARSE_CATALOG_PI were added.

An API function to install a custom character encoding converter is
now available. This makes it possible to use ICU for encoding conversion
even if libxml2 was compiled without ICU support, see example/icu.c.

### Deprecations

Access to many public struct members is now deprecated. Several accessor
functions were added to use instead.

More internal functions were deprecated.

### Removals

Metadata about the HTML4 content model was removed from the htmlElemDesc
struct and related functions were deprecated.

The FTP module and related functions were removed.

Support for the range and point extensions of the xpointer() scheme
was removed. The rest of the XPointer implementation isn't affected.
The xpointer() scheme now behaves like the xpath1() scheme.

Several legacy symbols and the functions in xmlunicode.h were removed.

Some unprefixed, internal macros like ATTRIBUTE_UNUSED were removed
from public headers.

ELF version information was removed.

The shell was moved from libxml2 to xmllint. Several related functions
are no longer available.

The libxml.m4 file containing autoconf macros was removed.

The --with-tree configuration option was removed.

The hack to detect single-threaded programs under glibc was removed.

### Planned removals

Support for HTTP and LZMA compression is planned to be removed in the
2.15 release.

The following features are considered for removal:


python/libxml.py
import libxml2mod

#
# Errors raised by the wrappers when some tree handling failed.
#
class treeError:
    def __init__(self, msg):
        self.msg = msg
    def __str__(self):
        return self.msg

class parserError:
    def __init__(self, msg):
        self.msg = msg
    def __str__(self):
        return self.msg

class uriError:
    def __init__(self, msg):
        self.msg = msg
    def __str__(self):
        return self.msg

class xpathError:
    def __init__(self, msg):
        self.msg = msg
    def __str__(self):
        return self.msg

class ioWrapper:
    def __init__(self, _obj):
        self.__io = _obj
        self._o = None

    def io_close(self):
        if self.__io == None:
	    return(-1)
	self.__io.close()
	self.__io = None
	return(0)

    def io_flush(self):
        if self.__io == None:
	    return(-1)
	self.__io.flush()
	return(0)

    def io_read(self, len = -1):
        if self.__io == None:
	    return(-1)
        if len < 0:
	    return(self.__io.read())
	return(self.__io.read(len))

    def io_write(self, str, len = -1):
        if self.__io == None:
	    return(-1)
        if len < 0:
	    return(self.__io.write(str))
	return(self.__io.write(str, len))

class ioReadWrapper(ioWrapper):
    def __init__(self, _obj, enc = ""):
        ioWrapper.__init__(self, _obj)
        self._o = libxml2mod.xmlCreateInputBuffer(self, enc)

    def __del__(self):
        print "__del__"
        self.io_close()
        if self._o != None:
            libxml2mod.xmlFreeParserInputBuffer(self._o)
        self._o = None

    def close(self):
        self.io_close()
        if self._o != None:
            libxml2mod.xmlFreeParserInputBuffer(self._o)
        self._o = None

class ioWriteWrapper(ioWrapper):
    def __init__(self, _obj, enc = ""):
        ioWrapper.__init__(self, _obj)
        self._o = libxml2mod.xmlCreateOutputBuffer(self, enc)

    def __del__(self):
        print "__del__"
        self.io_close()
        if self._o != None:
            libxml2mod.xmlOutputBufferClose(self._o)
        self._o = None

    def close(self):
        self.io_close()
        if self._o != None:
            libxml2mod.xmlOutputBufferClose(self._o)
        self._o = None

#
# Example of a class to handle SAX events
#
class SAXCallback:
    """Base class for SAX handlers"""
    def startDocument(self):
        """called at the start of the document"""
        pass

    def endDocument(self):
        """called at the end of the document"""
        pass

    def startElement(self, tag, attrs):
        """called at the start of every element, tag is the name of
	   the element, attrs is a dictionary of the element's attributes"""
        pass

    def endElement(self, tag):
        """called at the start of every element, tag is the name of
	   the element"""
        pass

    def characters(self, data):
        """called when character data have been read, data is the string
	   containing the data, multiple consecutive characters() callback
	   are possible."""
        pass

    def cdataBlock(self, data):
        """called when CDATA section have been read, data is the string
	   containing the data, multiple consecutive cdataBlock() callback
	   are possible."""
        pass

    def reference(self, name):
        """called when an entity reference has been found"""
        pass

    def ignorableWhitespace(self, data):
        """called when potentially ignorable white spaces have been found"""
        pass

    def processingInstruction(self, target, data):
        """called when a PI has been found, target contains the PI name and
	   data is the associated data in the PI"""
        pass

    def comment(self, content):
        """called when a comment has been found, content contains the comment"""
        pass

    def externalSubset(self, name, externalID, systemID):
        """called when a DOCTYPE declaration has been found, name is the
	   DTD name and externalID, systemID are the DTD public and system
	   identifier for that DTd if available"""
        pass

    def internalSubset(self, name, externalID, systemID):
        """called when a DOCTYPE declaration has been found, name is the
	   DTD name and externalID, systemID are the DTD public and system
	   identifier for that DTD if available"""
        pass

    def entityDecl(self, name, type, externalID, systemID, content):
        """called when an ENTITY declaration has been found, name is the
	   entity name and externalID, systemID are the entity public and
	   system identifier for that entity if available, type indicates
	   the entity type, and content reports it's string content"""
        pass

    def notationDecl(self, name, externalID, systemID):
        """called when an NOTATION declaration has been found, name is the
	   notation name and externalID, systemID are the notation public and
	   system identifier for that notation if available"""
        pass

    def attributeDecl(self, elem, name, type, defi, defaultValue, nameList):
        """called when an ATTRIBUTE definition has been found"""
	pass

    def elementDecl(self, name, type, content):
        """called when an ELEMENT definition has been found"""
	pass

    def entityDecl(self, name, publicId, systemID, notationName):
        """called when an unparsed ENTITY declaration has been found,
	   name is the entity name and publicId,, systemID are the entity
	   public and system identifier for that entity if available,
	   and notationName indicate the associated NOTATION"""
        pass

    def warning(self, msg):
        print msg

    def error(self, msg):
        raise parserError(msg)

    def fatalError(self, msg):
        raise parserError(msg)

#
# This class is the ancestor of all the Node classes. It provides
# the basic functionalities shared by all nodes (and handle
# gracefylly the exception), like name, navigation in the tree,
# doc reference, content access and serializing to a string or URI
#
class xmlCore:
    def __init__(self, _obj=None):
        if _obj != None: 
            self._o = _obj;
            return
        self._o = None

    def __getattr__(self, attr):
        if attr == "parent":
            ret = libxml2mod.parent(self._o)
            if ret == None:
                return None
            return xmlNode(_obj=ret)
        elif attr == "properties":
            ret = libxml2mod.properties(self._o)
            if ret == None:
                return None
            return xmlAttr(_obj=ret)
        elif attr == "children":
            ret = libxml2mod.children(self._o)
            if ret == None:
                return None
            return xmlNode(_obj=ret)
        elif attr == "last":
            ret = libxml2mod.last(self._o)
            if ret == None:
                return None
            return xmlNode(_obj=ret)
        elif attr == "next":
            ret = libxml2mod.next(self._o)
            if ret == None:
                return None
            return xmlNode(_obj=ret)
        elif attr == "prev":
            ret = libxml2mod.prev(self._o)
            if ret == None:
                return None
            return xmlNode(_obj=ret)
        elif attr == "content":
            return libxml2mod.xmlNodeGetContent(self._o)
        elif attr == "name":
            return libxml2mod.name(self._o)
        elif attr == "type":
            return libxml2mod.type(self._o)
        elif attr == "doc":
            ret = libxml2mod.doc(self._o)
            if ret == None:
		if self.type == "document_xml" or self.type == "document_html":
		    return xmlDoc(_obj=self._o)
		else:
		    return None
            return xmlDoc(_obj=ret)
        raise AttributeError,attr

        #
        # Those are common attributes to nearly all type of nodes
        #
    def get_parent(self):
        ret = libxml2mod.parent(self._o)
        if ret == None:
            return None
        return xmlNode(_obj=ret)
    def get_children(self):
        ret = libxml2mod.children(self._o)
        if ret == None:
            return None
        return xmlNode(_obj=ret)
    def get_last(self):
        ret = libxml2mod.last(self._o)
        if ret == None:
            return None
        return xmlNode(_obj=ret)
    def get_next(self):
        ret = libxml2mod.next(self._o)
        if ret == None:
            return None
        return xmlNode(_obj=ret)
    def get_properties(self):
        ret = libxml2mod.properties(self._o)
        if ret == None:
            return None
        return xmlAttr(_obj=ret)
    def get_doc(self):
        ret = libxml2mod.doc(self._o)
        if ret == None:
            return None
        return xmlDoc(_obj=ret)
    def get_prev(self):
        ret = libxml2mod.prev(self._o)
        if ret == None:
            return None
        return xmlNode(_obj=ret)
    def get_content(self):
        return libxml2mod.xmlNodeGetContent(self._o)
    def getContent(self):
        return libxml2mod.xmlNodeGetContent(self._o)
    def get_name(self):
        return libxml2mod.name(self._o)
    def get_type(self):
        return libxml2mod.type(self._o)
    def get_doc(self):
        ret = libxml2mod.doc(self._o)
        if ret == None:
            return None
        return xmlDoc(_obj=ret)
    def free(self):
        libxml2mod.freeDoc(self._o)

    #
    # Serialization routines, the optional arguments have the following
    # meaning:
    #     encoding: string to ask saving in a specific encoding
    #     format: if 1 the serializer is asked to indent the output
    #
    def serialize(self, encoding = None, format = 0):
        return libxml2mod.serializeNode(self._o, encoding, format)
    def saveTo(self, file, encoding = None, format = 0):
        return libxml2mod.saveNodeTo(self._o, file, encoding, format)
            
    #
    # Selecting nodes using XPath, a bit slow because the context
    # is allocated/freed every time but convenient.
    #
    def xpathEval(self, expr):
	doc = self.doc
	if doc == None:
	    return None
	ctxt = doc.xpathNewContext()
	ctxt.setContextNode(self)
	res = ctxt.xpathEval(expr)
	ctxt.xpathFreeContext()
	return res
	
#
# converters to present a nicer view of the XPath returns
#
def nodeWrap(o):
    # TODO try to cast to the most appropriate node class
    name = libxml2mod.name(o)
    if name == "element" or name == "text":
        return xmlNode(_obj=o)
    if name == "attribute":
        return xmlAttr(_obj=o)
    if name[0:8] == "document":
        return xmlDoc(_obj=o)
    if name[0:8] == "namespace":
        return xmlNs(_obj=o)
    if name == "elem_decl":
        return xmlElement(_obj=o)
    if name == "attribute_decl":
        return xmlAtribute(_obj=o)
    if name == "entity_decl":
        return xmlEntity(_obj=o)
    if name == "dtd":
        return xmlAttr(_obj=o)
    return xmlNode(_obj=o)

def xpathObjectRet(o):
    if type(o) == type([]) or type(o) == type(()):
        ret = map(lambda x: nodeWrap(x), o)
        return ret
    return o

#
# register an XPath function
#
def registerXPathFunction(ctxt, name, ns_uri, f):
    ret = libxml2mod.xmlRegisterXPathFunction(ctxt, name, ns_uri, f)

    
#
# For the xmlTextReader parser configuration
#
PARSER_LOADDTD=1
PARSER_DEFAULTATTRS=2
PARSER_VALIDATE=3

#
# Everything below this point is automatically generated
#

